adam_core/utils/indexable.py

```diff
@@ -1,174 +1,446 @@
+            ind = class_ind
 
-            elif isinstance(v, np.ndarray):
-                self.__dict__[k] = np.concatenate([self_v, v])
+        else:
+            raise IndexError("class_ind should be either an int or a slice.")
 
-            elif isinstance(v, Time):
-                self.__dict__[k] = Time(
-                    np.concatenate([self_v.mjd, v.mjd]), scale=v.scale, format="mjd"
+        # --- Check boundaries on the slice
+        if isinstance(ind, slice) and ind.start is not None and ind.start >= len(self):
+            raise IndexError(f"Index {ind.start} is out of bounds.")
+
+        if isinstance(ind, int) and self._class_index_to_members_is_slice:
+            member_ind = self._class_index_to_members[ind]
+
+        elif isinstance(ind, slice) and self._class_index_to_members_is_slice:
+
+            # Check if the array of slices are consecutive and share
+            # the same step size. If so, create a single slice that
+            # combines all of the slices.
+            slices = self._class_index_to_members[ind]
+            is_consecutive = True
+            for i, s_i in enumerate(slices[:-1]):
+                if s_i.stop != slices[i + 1].start:
+                    is_consecutive = False
+                    break
+                if s_i.step is not None and (s_i.step != slices[i + 1].step):
+                    is_consecutive = False
+                    break
+
+            if is_consecutive:
+                logger.debug(
+                    "Slices are consecutive and share the same step. "
+                    f"Combining slices a single slice with start {slices[0].start}, "
+                    f"end {slices[-1].stop} and step {slices[0].step}."
+                )
+                member_ind = slice(slices[0].start, slices[-1].stop, slices[0].step)
+            else:
+                logger.debug(
+                    "Slices are not consecutive. "
+                    "Combining slices a concatenating the members index for each slice."
+                )
+                member_ind = np.concatenate([self._class_index[s] for s in slices])
+
+        elif np.array_equal(self._class_index, self._member_index):
+            logger.debug("Using class index to index member arrays.")
+            member_ind = self._class_index[ind]
+        else:
+            logger.debug(
+                "Using unique class index to index member arrays with np.isin."
+            )
+            member_ind = self._member_index[
+                np.isin(
+                    self._class_index_to_members,
+                    self._class_index[ind],
                 )
+            ]
+        return member_ind
 
-            elif isinstance(v, (list, Indexable)):
-                self_v.append(v)
+    def __len__(self):
+        if self._class_index is not None:
+            return len(self._class_index)
+        else:
+            err = "Length is not defined for this class."
```

Comment:
@spenczar Ah, great catch! So this was originally added to make sure that I remembered to initialize the Indexable class in its subclasses. As you point out, because the constructor calls `set_index` the length is always defined but if you subclass and forget the call to `super(self).__init__` then it won't work. Is this reasonable behavior? Should I modify the error to state that the Indexable was not properly initialized or just remove it and assume that the person writing the subclass knows to call the constructor?

Reply:
I'd just remove it. When subclassing, you should **always** call the parent class's `__init__` method unless you have an extremely good reason not to.

Reply:
Removed!